src/hotspot/share/runtime/os.cpp

```diff
@@ -2208,9 +2208,9 @@ bool os::release_memory(char* addr, size_t bytes) {
   assert_nonempty_range(addr, bytes);
   bool res;
   if (MemTracker::enabled()) {
-    ThreadCritical tc;
     res = pd_release_memory(addr, bytes);
     if (res) {
+      ThreadCritical tc;
```

Comment:
What is the reason for moving the memory operation out of the protection of the lock? Wouldn't it be better if the memory operation and the accounting were done together atomically like with [`os::uncommit_memory`](https://github.com/openjdk/jdk/pull/20425/files#diff-c626d392182b0193ad0e754de3f4fd42c98d959edade67490b6e1f24be38efdbR2189) for example.

Reply:
TBH, I don't know and this is not a change in this PR. The uncommit_memory has the same approach. I just know that if we move the pd_xxxx inside the lock some tests fail or some assertions happen, but do not remember which and why.

Reply:
OK, I guess it doesn't make much of a difference at the moment since it's inconsistent across many memory operations anyway. Commit, uncommit, reserve, and release don't agree on whether to make the memory operation and NMT accounting atomic together. I opened an issue for this here: https://bugs.openjdk.org/browse/JDK-8341491